### PR TITLE
Mark file resource functions as having side effects

### PR DESCRIPTION
**Pull request by jlherren**
Fixes phpstan/phpstan#5461

### CODE REVIEW

File: resources/functionMetadata.php [hunk 1]
```diff
@@ -837,13 +844,25 @@
...
+	'fseek' => ['hasSideEffects' => true],
+	'fstat' => ['hasSideEffects' => true],
+	'ftell' => ['hasSideEffects' => true],
 	'ftok' => ['hasSideEffects' => false],
+	'ftruncate' => ['hasSideEffects' => true],
 	'func_get_arg' => ['hasSideEffects' => false],
 	'func_get_args' => ['hasSideEffects' => false],
 	'func_num_args' => ['hasSideEffects' => false],
 	'function_exists' => ['hasSideEffects' => false],
+	'fwrite' => ['hasSideEffects' => true],
```
Comment (staabm) [inline resources/functionMetadata.php:865]: this reminds me of, whether `file_put_contents` and `file_get_contents` is missing in this list?

Comment (jlherren) [reply resources/functionMetadata.php:865]: I only included the functions that work on file handles. But I guess all filesystem operations probably should be added. In that case it should also include `unlink()`, `file()`, `mkdir()`, and tons of others. I'll try to do that and see if anything breaks.

**Comment by jlherren** [comment]
Okay, here's a revised version, which adds most [filesystem functions](https://www.php.net/manual/en/ref.filesystem.php). `file_get_content()` is now marked to not have side-effects, which kinda makes sense. However, I'm still a bit unsure about how useful that is. It makes this work though:

```
if (file_get_contents($path) === 'data') {
	assertType('\'data\'', file_get_contents($path));
	file_put_contents($path, 'other');
	assertType('string|false', file_get_contents($path));
}
```

Maybe in the future there should be more precise metadata than just "hasSideEffects". Some of these functions would be better described as "volatile". Or perhaps even an interface `FunctionSideEffectExtension` or something like that.

**Comment by ondrejmirtes** [comment]
Hi, thank you! The file you modified is actually generated automatically when running `./bin/generate-function-metadata.php` and the result is merged from `./bin/functionMetadata_original.php` and by looking at jetbrains/phpstorm-stubs methods and functions with the `#[Pure]` attribute.

I'd like this course of action instead:

1) Contribute the functions that have `'hasSideEffects' => true` as a modification to `functionMetadata_original.php`.
2) Contribute the `#[Pure]` functions without side effects to `jetbrains/phpstorm-stubs`.
3) Once the PR from 2) is merged, please update the package here and run `./bin/generate-function-metadata.php`.

Thank you!

**Comment by jlherren** [comment]
This got a bit complicated unfortunately.

The [phpstorm-stubs update](https://github.com/JetBrains/phpstorm-stubs/pull/1244) has been merged. However, updating phpstorm-stubs to the newest master causes a few tests to fail. This is due to a [change in phpstorm-stubs](https://github.com/JetBrains/phpstorm-stubs/pull/1234/files), which implements [PHP RFC: Add return type declarations for internal methods](https://wiki.php.net/rfc/internal_method_return_types). That RFC is about adding tentative return type declarations, which as I understand is about return type declarations that only trigger a warning, but aren't fully enforced yet (not until PHP 9.0).

As an example, `Countable::count()` has now a tentative return type `int`, causing PHPStan to complain about all implementations that don't have a compatible return type declaration.

I'm not sure what the best course of action here. In the long term, PHPStan should probably parse `#[TentativeType]` and act on it. For now I'll reduce this PR to only update `./bin/functionMetadata_original.php`, it should be enough to fix phpstan/phpstan#5461.

**Comment by jlherren** [comment]
I see that tentative return types have been implemented and the stubs have been updated. I rebased this to fix conflicts.

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**